language/vm/vm-runtime/src/move_vm.rs

```diff
@@ -72,17 +90,148 @@ impl VMExecutor for MoveVM {
             // needs to be fixed.
             VMRuntime::new(&*arena, config)
         });
-        vm.rent(|runtime| runtime.execute_block_transactions(transactions, state_view))
+
+        let mut result = vec![];
+        let blocks = chunk_block_transactions(transactions);
+        let mut data_cache = BlockDataCache::new(state_view);
+        for block in blocks {
+            match block {
+                TransactionBlock::UserTransaction(txns) => {
+                    let outs = vm.rent(|runtime| {
+                        execute_user_transaction_block(txns, runtime, &mut data_cache, state_view)
+                    });
+                    result.append(&mut outs?);
+                }
+                TransactionBlock::BlockPrologue(block_metadata) => {
+                    result.push(vm.rent(|runtime| {
+                        process_block_metadata(block_metadata, runtime, state_view, &mut data_cache)
+                    }))
+                }
+                // TODO: Implement the logic for processing writeset transactions.
+                TransactionBlock::WriteSet(_) => unimplemented!(""),
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+/// Transactions divided by transaction flow.
+/// Transaction flows are different across different types of transactions.
+pub enum TransactionBlock {
+    UserTransaction(Vec<SignedTransaction>),
+    WriteSet(WriteSet),
+    BlockPrologue(BlockMetadata),
+}
+
+pub fn chunk_block_transactions(txns: Vec<Transaction>) -> Vec<TransactionBlock> {
```

Comment:
this is only public because of testing but it seems we need a better way to plug in tests and understand the dependency graph